File: mvd/mvd/report/email_an_coop_senden/email_an_coop_senden.py
# Copyright (c) 2013, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
    columns = [
        {"label": _("RSV Mitglied"), "fieldname": "name", "fieldtype": "Link", "options": "RSVMitglied", "width": 180},
        { "label": _("Mitglied-Nr."), "fieldname": "mitglied_nr", "fieldtype": "Data", "width": 110},
        { "label": _("Vorname"), "fieldname": "vorname", "fieldtype": "Data", "width": 120},
        { "label": _("Nachname"), "fieldname": "nachname", "fieldtype": "Data", "width": 120},
        { "label": _("RSV Mandat"), "fieldname": "rsvmandat", "fieldtype": "Link", "options": "RSVMandat", "width": 150},
        { "label": _("Anwält*in"), "fieldname": "anwalt", "fieldtype": "Link", "options": "Termin Kontaktperson", "width": 160},
        { "label": _("Aktion"), "fieldname": "action_button", "fieldtype": "Data", "width": 180}
    ]

    data = frappe.db.sql("""
        SELECT 
            m.name,
            m.mitglied_nr,
            m.vorname,
            m.nachname,
            m.rsvmandat,
            rm.anwalt
        FROM 
            `tabRSVMitglied` m
        INNER JOIN 
            `tabRSVMandat` rm ON m.rsvmandat = rm.name
        WHERE 
            (m.sendung_an_coop = 0 OR m.sendung_an_coop IS NULL)
            AND rm.anwalt IS NOT NULL 
            AND rm.anwalt != ''
        ORDER BY 
            m.modified DESC
    """, as_dict=True)

    for row in data:
        row["action_button"] = f'''<button class="btn btn-xs btn-primary btn-send-coop-email" data-name="{row["name"]}">
            <i class="fa fa-envelope"></i> E-Mail an Coop senden
        </button>'''

    return columns, data

File: mvd/mvd/report/email_an_coop_senden/email_an_coop_senden.js
// Copyright (c) 2016, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Email an Coop senden"] = {
"filters": [
    ],
    "onload": function(report) {
        report.page.main.on("click", ".btn-send-coop-email", function() {
            let docname = $(this).attr("data-name");

            frappe.call({
                method: "mvd.mvd.doctype.rsvmitglied.rsvmitglied.get_coop_email_data",
                args: { "docname": docname },
                freeze: true,
                freeze_message: __('E-Mail-Daten werden geladen (ZIP wird erstellt)...'),
                callback: function(r) {
                    if (!r.exc && r.message) {
                        var mail_data = r.message;

                        let d = new frappe.ui.Dialog({
                            title: __('E-Mail an Coop: ' + docname),
                            size: 'large',
                            fields: [
                                {fieldtype: 'Data', fieldname: 'recipients', label: 'An', reqd: 1, default: mail_data.recipients},
                                {fieldtype: 'Data', fieldname: 'cc', label: 'CC', default: mail_data.cc},
                                {fieldtype: 'Data', fieldname: 'subject', label: 'Betreff', reqd: 1, default: mail_data.subject},
                                {fieldtype: 'Text Editor', fieldname: 'content', label: 'Nachricht', reqd: 1, default: mail_data.content}
                            ],
                            primary_action_label: __('Senden'),
                            primary_action: function(values) {
                                d.get_primary_btn().prop('disabled', true);

                                frappe.call({
                                    method: "mvd.mvd.doctype.rsvmitglied.rsvmitglied.send_coop_email_custom",
                                    args: {
                                        docname: docname,
                                        recipients: values.recipients,
                                        cc: values.cc,
                                        subject: values.subject,
                                        content: values.content
                                    },
                                    freeze: true,
                                    freeze_message: __('E-Mail wird versendet...'),
                                    callback: function(send_r) {
                                        if (!send_r.exc) {
                                            frappe.show_alert({
                                                message: __('E-Mail für {0} erfolgreich gesendet.', [docname]), 
                                                indicator: 'green'
                                            });
                                            d.hide();
                                            report.refresh();
                                        } else {
                                            d.get_primary_btn().prop('disabled', false);
                                        }
                                    }
                                });
                            }
                        });
                        d.show();
                    }
                }
            });
        });
    }
};